pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	resourceType     = "resource_type"
+	nodeName         = "name"
+	nodeProvisioner  = "provisioner"
+	nodeZone         = "zone"
+	nodeArchitecture = "arch"
+	nodeCapacityType = "capacity_type"
+	nodeInstanceType = "instance_type"
+	nodePhase        = "phase"
+	nodeLabels       = "node_labels"
+)
+
+var (
+	allocatableGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "allocatable",
+			Help:      "Node allocatable",
+		},
+		getLabelNames(),
+	)
+	requestsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_requests",
+			Help:      "Node total requests",
+		},
+		getLabelNames(),
+	)
+	limitsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_limits",
+			Help:      "Node total limits",
+		},
+		getLabelNames(),
+	)
+	overheadGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_overhead",
+			Help:      "Node total overhead",
+		},
+		getLabelNames(),
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(allocatableGaugeVec)
+	crmetrics.Registry.MustRegister(requestsGaugeVec)
+	crmetrics.Registry.MustRegister(limitsGaugeVec)
+	crmetrics.Registry.MustRegister(overheadGaugeVec)
+}
+
+func getLabelNames() []string {
+	return []string{
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodeZone,
+		nodeArchitecture,
+		nodeCapacityType,
+		nodeInstanceType,
+		nodePhase,
+		nodeLabels,
+	}
+}
+
+type Controller struct {
+	KubeClient client.Client
+	LabelsMap  map[types.NamespacedName][]prometheus.Labels
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		KubeClient: kubeClient,
+		LabelsMap:  make(map[types.NamespacedName][]prometheus.Labels),
+	}
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("nodemetrics").With("node", req.Name))
+	// Retrieve node from reconcile request
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to node deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				c.deleteGauges(labels)
```

Comment:
It looks like we're deleting in all cases except wen GetNode returns an error. I'd argue that we might actually want to clean up if we're unable to reconcile for some reason, since we don't know what the data is. It may be simpler to just reset the data as the first step of the reconcile loop.

Reply:
done